<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
  <head><script type="text/javascript" src="/static/js/analytics.js?v=1493139128.0" charset="utf-8"></script>

<script type="text/javascript">archive_analytics.values.service='wb';archive_analytics.values.server_name='wwwb-app40.us.archive.org';archive_analytics.values.server_ms=448;</script>
<script type="text/javascript" src="/static/js/wbhack.js?v=1493139128.0" charset="utf-8"></script>

<script type="text/javascript">
__wbhack.init('');
</script>
<link rel="stylesheet" type="text/css" href="/static/css/banner-styles.css?v=1493139128.0" />
<link rel="stylesheet" type="text/css" href="/static/css/iconochive.css?v=1493139128.0" />

<!-- End Wayback Rewrite JS Include --><link rel="icon" type="image/vnd.microsoft.icon" href="/web/20100625030629im_/http://mor.clarku.edu/images/favicon.ico">
    <link rel="stylesheet" href="/web/20100625030629cs_/http://mor.clarku.edu/css/mainstyle.css" type="text/css"/>
    <meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1"/>
    	
		
    <title>mor -Scripts</title>
  </head>
  <body><!-- BEGIN WAYBACK TOOLBAR INSERT -->
<script type="text/javascript" src="/static/js/timestamp.js?v=1493139128.0" charset="utf-8"></script>
<script type="text/javascript" src="/static/js/graph-calc.js?v=1493139128.0" charset="utf-8"></script>
<script type="text/javascript" src="/static/js/auto-complete.js?v=1493139128.0" charset="utf-8"></script>
<script type="text/javascript" src="/static/js/toolbar.js?v=1493139128.0" charset="utf-8"></script>

<style type="text/css">
body {
  margin-top:0 !important;
  padding-top:0 !important;
  min-width:800px !important;
}
.wb-autocomplete-suggestions {
    text-align: left; cursor: default; border: 1px solid #ccc; border-top: 0; background: #fff; box-shadow: -1px 1px 3px rgba(0,0,0,.1);
    position: absolute; display: none; z-index: 2147483647; max-height: 254px; overflow: hidden; overflow-y: auto; box-sizing: border-box;
}
.wb-autocomplete-suggestion { position: relative; padding: 0 .6em; line-height: 23px; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; font-size: 1.02em; color: #333; }
.wb-autocomplete-suggestion b { font-weight: bold; }
.wb-autocomplete-suggestion.selected { background: #f0f0f0; }
</style>
<div id="wm-ipp" lang="en" style="display:none;direction:ltr;">
<div style="position:fixed;left:0;top:0;width:100%!important">
<div id="wm-ipp-inside">
  <table style="width:100%;">
    <tbody>
      <tr>
	<td id="wm-logo">
	  <a href="/web/" title="Wayback Machine home page"><img src="/static/images/toolbar/wayback-toolbar-logo.png" alt="Wayback Machine" width="110" height="39" border="0" /></a>
	</td>
	<td class="c">
	  <table style="margin:0 auto;">
	    <tbody>
	      <tr>
		<td class="u" colspan="2">
		  <form target="_top" method="get" action="/web/submit" name="wmtb" id="wmtb"><input type="text" name="url" id="wmtbURL" value="http://mor.clarku.edu/index.php?mor=morScripts" style="width:400px;" onfocus="this.focus();this.select();" /><input type="hidden" name="type" value="replay" /><input type="hidden" name="date" value="20100625030629" /><input type="submit" value="Go" /></form>
		</td>
		<td class="n" rowspan="2">
		  <table>
		    <tbody>
		      <!-- NEXT/PREV MONTH NAV AND MONTH INDICATOR -->
		      <tr class="m">
			<td class="b" nowrap="nowrap">May</td>
			<td class="c" id="displayMonthEl" title="You are here: 03:06:29 Jun 25, 2010">JUN</td>
			<td class="f" nowrap="nowrap">Jul</td>
		      </tr>
		      <!-- NEXT/PREV CAPTURE NAV AND DAY OF MONTH INDICATOR -->
		      <tr class="d">
			<td class="b" nowrap="nowrap"><img src="/static/images/toolbar/wm_tb_prv_off.png" alt="Previous capture" width="14" height="16" border="0" /></td>
			<td class="c" id="displayDayEl" style="width:34px;font-size:24px;" title="You are here: 03:06:29 Jun 25, 2010">25</td>
			<td class="f" nowrap="nowrap"><img src="/static/images/toolbar/wm_tb_nxt_off.png" alt="Next capture" width="14" height="16" border="0" /></td>
		      </tr>
		      <!-- NEXT/PREV YEAR NAV AND YEAR INDICATOR -->
		      <tr class="y">
			<td class="b" nowrap="nowrap">2009</td>
			<td class="c" id="displayYearEl" title="You are here: 03:06:29 Jun 25, 2010">2010</td>
			<td class="f" nowrap="nowrap">2011</td>
		      </tr>
		    </tbody>
		  </table>
		</td>
	      </tr>
	      <tr>
		<td class="s">
		  		  <div id="wm-nav-captures">
		    		    <a class="t" href="/web/20100625030629*/http://mor.clarku.edu/index.php?mor=morScripts" title="See a list of every capture for this URL">1 capture</a>
		    <div class="r" title="Timespan for captures of this URL">25 Jun 2010</div>
		    </div>
		</td>
		<td class="k">
		  <a href="" id="wm-graph-anchor">
		    <div id="wm-ipp-sparkline" title="Explore captures for this URL" style="position: relative">
		      <canvas id="wm-sparkline-canvas" width="550" height="27" border="0"></canvas>
		    </div>
		  </a>
		</td>
	      </tr>
	    </tbody>
	  </table>
	</td>
	<td class="r">
	  <a href="#" onclick="window.open('https://www.facebook.com/sharer/sharer.php?u=https://web-beta.archive.org/web/20100625030629/http://mor.clarku.edu/index.php?mor=morScripts', '', 'height=400,width=600'); return false;" title="Share on Facebook" style="top: 4px; right: 16px;" target="_blank"><span class="iconochive-facebook" style="color:#3b5998;font-size:160%;"></span></a>
	  <a href="#" onclick="window.open('https://twitter.com/intent/tweet?text=https://web-beta.archive.org/web/20100625030629/http://mor.clarku.edu/index.php?mor=morScripts&amp;via=internetarchive', '', 'height=400,width=600'); return false;" title="Share on Twitter" style="bottom:14px;right:16px;" target="_blank"><span class="iconochive-twitter" style="color:#1dcaff;font-size:160%;"></span></a>
	  <a href="http://faq.web.archive.org/" title="Get some help using the Wayback Machine" style="bottom:14px;right:5px;padding-right:0;"><span class="iconochive-question" style="color:rgb(87,186,244);font-size:160%;"></span></a>
	  <a href="#close" onclick="__wm.h(event);return false;" style="top:-2px;right:0;padding-right:0;" title="Close the toolbar"><span class="iconochive-remove-circle" style="color:#888888;font-size:240%;"></span></a>
	  <a id="wm-expand" class="wm-btn" href="#expand" onclick="__wm.ex(event);return false;"><span class="iconochive-down-solid"></span> <span style="font-size:80%">About this capture</span></a>
	</td>
      </tr>
    </tbody>
  </table>
    <div id="wm-capinfo" style="border-top:1px solid #777;display:none;">
            <div style="background-color:#353535;color:#aaa;font-weight:bold;text-align:center;"><a class="wm-selector selected" href="javascript:void(0)">COLLECTED BY</a></div>
    <div style="padding:3px;position:relative;">
            <div style="display:inline-block;vertical-align:top;width:50%;">
			<span class="c-logo" style="background-image:url(https://archive.org/services/img/alexacrawls);"></span>
		Organization: <a style="color:#33f;" href="https://archive.org/details/alexacrawls" target="_new"><span class="wm-title">Alexa Crawls</span></a>
		<div style="max-height:75px;overflow:hidden;position:relative;">
	  <div style="position:absolute;top:0;left:0;width:100%;height:75px;background:linear-gradient(to bottom,rgba(255,255,255,0) 0%,rgba(255,255,255,0) 90%,rgba(255,255,255,255) 100%);"></div>
	  Starting in 1996, <a href="http://www.alexa.com/">Alexa Internet</a> has been donating their crawl data to the Internet Archive.  Flowing in every day, these data are added to the <a href="http://web.archive.org/">Wayback Machine</a> after an embargo period.
	</div>
	      </div>
      <div style="display:inline-block;vertical-align:top;width:49%;">
			<span class="c-logo" style="background-image:url(https://archive.org/services/img/alexacrawls)"></span>
		<div>Collection: <a style="color:#33f;" href="https://archive.org/details/alexacrawls" target="_new"><span class="wm-title">Alexa Crawls</span></a></div>
		<div style="max-height:75px;overflow:hidden;position:relative;">
	  <div style="position:absolute;top:0;left:0;width:100%;height:75px;background:linear-gradient(to bottom,rgba(255,255,255,0) 0%,rgba(255,255,255,0) 90%,rgba(255,255,255,255) 100%);"></div>
	  Starting in 1996, <a href="http://www.alexa.com/">Alexa Internet</a> has been donating their crawl data to the Internet Archive.  Flowing in every day, these data are added to the <a href="http://web.archive.org/">Wayback Machine</a> after an embargo period.
	</div>
	      </div>
    </div></div></div></div></div><script type="text/javascript">
__wm.bt(550,27,25,2,"web","http://mor.clarku.edu/index.php?mor=morScripts","2010-06-25",1996);
</script>
<!-- END WAYBACK TOOLBAR INSERT -->
    <div id="header"><!--header-->
      <a href="/web/20100625030629/http://mor.clarku.edu/index.php?mor=home"><img id="logo" src="/web/20100625030629im_/http://mor.clarku.edu/images/mor_logo.gif" alt="Mor"/></a>
      <h3 class="loginForm title">Cladesys Login</h3>
<form action="index.php?mor=login" method="post" class="loginForm">
    <p>email:</p>
    <input type="text" name="email" maxlength="40" size="15"/><br/>
    <p>password:</p>
    <input type="password" name="password" maxlength="30" size="15"/><br/>
    <input type="submit" name="doLogin" value="log in"/>
</form>
    </div><!--header end-->
    <div id="body"><!--body-->
      <div class="leftMenu">
          <h3 class="leftMenu header">explore</h3>
          <ul class="leftMenu list">
            <li><a href="/web/20100625030629/http://mor.clarku.edu/index.php?mor=home">morHome</a></li>
            <li><a href="/web/20100625030629/http://mor.clarku.edu/index.php?mor=flowChart">Documentation</a></li>
            <li><a href="/web/20100625030629/http://mor.clarku.edu/index.php?mor=download">Download</a></li>
          </ul>
          <h3 class="leftMenu header">trees</h3>
          <ul class="leftMenu list">
            <li><a href="/web/20100625030629/http://mor.clarku.edu/mp.html">Parsimony</a></li>
            <li><a href="/web/20100625030629/http://mor.clarku.edu/nj.html">Jackknife</a></li>
            <li><a href="/web/20100625030629/http://mor.clarku.edu/ml.html">Maximum Likelihood</a></li>
            <li><a href="/web/20100625030629/http://mor.clarku.edu/backbone.html">Backbone Constraint</a></li>
            <li><a href="/web/20100625030629/http://mor.clarku.edu/index.php?mor=archive">Archives</a></li>
            <!--<li><a href="" class="error">Recent Changes</a></li>-->
            <li><a href="/web/20100625030629/http://mor.clarku.edu/index.php?mor=treeViz">Tree Visualization</a></li>
          </ul>
          <h3 class="leftMenu header">sequences</h3>
          <ul class="leftMenu list">
            <li><a href="/web/20100625030629/http://mor.clarku.edu/index.php?mor=accepted">Current LSU rRNA sequences</a></li>
            <li><a href="/web/20100625030629/http://mor.clarku.edu/index.php?mor=rejected">Rejected LSU rRNA sequences</a></li>
            <li><a href="/web/20100625030629/http://mor.clarku.edu/index.php?mor=tansley">ITS sequence clusters</a></li>
          </ul>
          <h3 class="leftMenu header">interact</h3>
          <ul class="leftMenu list">
            <!--<li><a href="" class="error">Comments</a></li>-->
            <li><a href="/web/20100625030629/http://mor.clarku.edu/index.php?mor=register">Register</a></li>
            <li><a href="/web/20100625030629/http://mor.clarku.edu/index.php?mor=morblast">morBlast</a></li>
          </ul>
          <h3 class="leftMenu header">miscellaneous</h3>
          <ul class="leftMenu list">
            <li><a href="/web/20100625030629/http://mor.clarku.edu/index.php?mor=links">morLinks</a></li>
          </ul>
          <h3 class="leftMenu header">development</h3>
          <ul class="leftMenu list">
            <li><a href="/web/20100625030629/http://mor.clarku.edu/index.php?mor=morTeam">morTeam</a></li>
            <li><a href="/web/20100625030629/http://mor.clarku.edu/index.php?mor=blog">Dev Blog</a></li>
          </ul>
      </div><!--end left menu-->
      <div id="content">
        <div class="home">
<h2 class="pageTitle">mor Scripts</h2><br/>

<ul class="morDoc">
        <li><a href="#Script1">mor_download.pl</a> <br>Downloads, screens and aligns the LSU data.</li>
        <li><a href="#Script2">mor_analysis.pl</a> <br>Creates a ML, MP and NJ tree, uploads them to the website and parses them for clades.</li>
	<li><a href="#Script3">cluster_creator.pl</a> <br>Downloads and groups ITS sequences into clusters based on similarity.</li>
        <li><a href="#Script4">cluster_stats.pl</a> <br>Summarizes the data from cluster_creator.pl into a table for easier display online.</li>
        <li><a href="#Script5">mor_matches.pl</a> <br>Makes matches between the LSU and ITS data then it saves to a table.</li>
</ul>

</p>



<p><a name="Script1"></a></p>
<h4 class="sectionTitle">mor_download.pl</h4>
<p> 
This script accomplishes three fairly straight forward tasks that account for organizing and generating the sequence level data that will be used in mor_analysis.pl
</p><p>
It firsts downloads the sequences from genbank. Using the query phrase supplied in the setting file and the date of the most recently submitted sequence, it connects to genbank to download all of the sequences that match the criteria. Using bioperl's ability to download batches of sequences, it is able to accomplish this task at a much faster rate than downloading single sequences at a time due to genbank's 3 second delay between downloading files. This is not without its costs though, genbank's high level of traffic result in the occasional timeout which prevents mor from downloading all the sequences at the same time (however the latest version of mor should rectify that with it's download function). All of these sequences are then written to a standard genbank flat file which is written over in every run and stored in the mor database (this includes some of the feature information as well as the unaligned sequence.
</p><p>
The next step is screening the sequences to make sure they match our definition of what a homobasidiomycete's LSU sequence should look like. Using a hidden markov model as implemented in hmmer from EMBOSS, we screen out all of the sequences that do not resemble the profile of a homobasidiomycete. The creation of an accurate profile is the more difficult and most important step of this process, without an accurate profile even sequences in the backbone will be rejected. We also have implemented a length requirement (which is also tested for in the query phrase) and a quality requirement of a limited number of ambiguous nucleotides. The sequences are then marked as accepted or rejected with their reasons for rejection stored in the database.
</p><p>
The script then retrieves all of the sequences that have been marked as accepted from the mor database and aligns them using mafft. After which the aligned sequences are then stored in the database for use with mor_analysis.pl.
</p><p>
A HRI (Hibbett Redundancy Index), which weeds out the sequences more than 99.5 percent identical, was used to reduce the number of sequences for analysis, however this didn't remove enough sequences to warrant reimplementation in mor 2.0. Future versions of mor may do well to reimplement this function as more redundant sequences are deposited into genbank. 
</p>

<p><a name="Script2"></a></p>
<h4 class="sectionTitle">mor_analysis.pl</h4>
<p> 
This portion of mor deals mostly with generating and manipulating trees. It uses the aligned sequences from the mor database to generate three trees: a maximum likelihood and a maximum parsimony , a neighbor-joining jackknife tree. All three trees make use of a backbone tree to constrain the tree and speed up the process
</p><p>
The maximum likelihood tree is the simplest for the program to construct, it simply used a pre-determined command to run the analysis on the alignment that was generated for that run. It uses a multi-threaded version of raxml to do this. After the analysis is done, the tree is saved to a tree database for further use and some more processing later.
</p><p>
The next two trees are generated using PAUP* in a slightly more complicated way. Using keywords from the settings file, a template for a nexus paup block is modified to suit the current run. It then pre-pends the alignment and runs paup. The neighbor-joining tree, although individually takes the shortest amount of time to run, must run 100 times to generate the confidence intervals for the tree, which means it takes the longest. The maximum parsimony tree is set to run for five days (according to a line in the settings file).
</p><p>
After all three trees are generated, we use the maximum parsimony tree to generate user-defined clades via taxon based clade definitions. These clade trees are saved to their own table.
</p><p>
The final step in the process amounts to turning the trees generated in the process into something that a person can read and interact with on the website. We settled on a simple html tree that is generated using the output of paup*. 
</p>

<p><a name="Script3"></a></p>
<h4 class="sectionTitle">cluster_creator.pl</h4>
<p> 
This section of the scripts is slightly divergent from the mor_* scripts. This script's goal is simply to download, group similar sequences together and then align those groups. That said, how it goes about grouping the sequences is slightly complicated and needs some explaining.
</p><p>
Before cluster_creator.pl gets to creating clusters, it first has to download and screen the sequences. It does this in a very similar way to <a href="#Script1">mor_download.pl</a> does only without the hidden markov model screen.
</p><p>
The next step is clustering, which consists of three steps. To cluster the sequences, a group of sequences is identified in the data base and cluster_creator attempts to match them based on their percent sequence similarity to a reference sequence for a particular cluster. If that sequence can not be matched, it becomes its own reference to its own group. However, this is a challenge due to the fact that no matter how you slice it, clustering will (probably) never be O(n) so we have to find ways of reducing the size of n. This is why the first step in cluster_creator.pl uses the genbank taxonomy associated with each sequence. Using non-nested classifiers to pre-cluster the sequences, cluster_creator is able to run through several smaller groups at a much faster rate than running through one large group. The second step is to check the reference sequences against each other to make sure that more clusters haven't been artificially created due to a misidenfication. If it detects a reference sequence that should have been attached to another cluster, it dissolves that cluster and puts all the sequences associated with the bad reference back into a pile for future clustering. The final step is to simply cluster the remaining sequences that were either part of a bad cluster or simply hadn't been pre-clustered in the first step.
</p><p>
The last step is to make alignments of each of the clusters and store the aligned sequences back into the table. It is best to think of this table as a combination of Bio::Seq and Bio::Align objects. However, the Bio::Align objects aren't the easiest to display online so <a href="#Script4">cluster_stats.pl</a> steps in to address that issue. 
</p>

<p><a name="Script4"></a></p>
<h4 class="sectionTitle">cluster_stats.pl</h4>
<p> 
This portion of mor is relatively straight forward. It simply goes through all of the clusters and collects information about each of the clusters to store in a separate table to display online. It collects the date of the most recently added sequence, the size, the average similarity in the cluster and what pre-cluster it was initially in. 
</p>

<p><a name="Script5"></a></p>
<h4 class="sectionTitle">mor_matches.pl</h4>
<p> 
This final script makes use of the feature data present in the mor and its tables. It is able to make two levels of matches for each feature common between the two tables. The first level is what we call a "loose" match, which is simply if all the letters and numbers match and are in the same order in both tables (eg. DSH-10.1 would match DSH_10.1). The second level is built on the loose matches. If mor_matches detects a loose match, it then checks to see if the features strictly match (DSH-10.1 wouldn't match DSH_10.1). One can see the obvious strengths and limitations to this approach, which is why the matches table has a field to allow for the match to be human broken/made. 
</p>

</div>

      </div>
      <div id="footer">
        <span class="mor">mor</span> is being built by <a href="https://web-beta.archive.org/web/20100625030629/mailto:dhibbett@clarku.edu?subject=mor">Hibbett D</a>
        , Nilsson RH, Shonfeld M, Snyder M, Costanzo J, Fonseca M, Twomey R, Gaytan B, Stein P, Burke JP, Heider T, Ha C and Ohman A
      </div>
    </div><!--end body-->
  </body>
</html>
